src/python/pants/backend/shell/shunit2_test_runner.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import re
+from dataclasses import dataclass
+
+from pants.backend.shell.shell_setup import ShellSetup
+from pants.backend.shell.target_types import (
+    ShellSources,
+    Shunit2Shell,
+    Shunit2ShellField,
+    Shunit2Tests,
+    Shunit2TestsSources,
+    Shunit2TestsTimeout,
+)
+from pants.core.goals.test import (
+    TestDebugRequest,
+    TestExtraEnv,
+    TestFieldSet,
+    TestResult,
+    TestSubsystem,
+)
+from pants.core.target_types import FilesSources, ResourcesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.environment import Environment, EnvironmentRequest
+from pants.engine.fs import (
+    CreateDigest,
+    Digest,
+    DigestContents,
+    DownloadFile,
+    FileContent,
+    FileDigest,
+    MergeDigests,
+)
+from pants.engine.process import (
+    BinaryNotFoundError,
+    BinaryPath,
+    BinaryPathRequest,
+    BinaryPaths,
+    FallibleProcessResult,
+    InteractiveProcess,
+    Process,
+    ProcessCacheScope,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import Sources, TransitiveTargets, TransitiveTargetsRequest
+from pants.engine.unions import UnionRule
+from pants.option.global_options import GlobalOptions
+from pants.util.logging import LogLevel
+from pants.util.strutil import create_path_env_var
+
+
+@dataclass(frozen=True)
+class Shunit2FieldSet(TestFieldSet):
+    required_fields = (Shunit2TestsSources,)
+
+    sources: Shunit2TestsSources
+    timeout: Shunit2TestsTimeout
+    shell: Shunit2ShellField
+
+
+@dataclass(frozen=True)
+class TestSetupRequest:
+    field_set: Shunit2FieldSet
+
+
+@dataclass(frozen=True)
+class TestSetup:
+    process: Process
+
+
+class Shunit2AlreadySourced(Exception):
+    pass
+
+
+class ShellNotConfigured(Exception):
+    pass
+
+
+def validate_source_shunit2_not_included(fc: FileContent, address: Address) -> None:
+    lines = fc.content.splitlines()
+    for i, line in enumerate(lines):
+        if re.search(rb"(?:source|\.) +[.${}/'\"\w]*shunit2\b", line):
+            raise Shunit2AlreadySourced(
+                f"The test file {fc.path} sources shunit2 on line {i + 1} with: "
```

Comment:
This seems like a pretty suboptimal UX. Why can't we just materialize their file with this line replaced so that they can continue to use their test outside of pants? If its possible, I think we should always strive to leave a working system undisturbed. We do this for pytest tests afaict for example.

Reply:
Good point. Will fix.